metagamer/environments/qagent.py

```diff
@@ -1,144 +1,432 @@
+
+
+        Returns: a tuple representing an unfilled coordinate that's selected.
+
+        """
+        for row in range(state.shape[0]):
+            for col in range(state.shape[1]):
+                if state[row, col] == 0:
+                    return row, col
+
+    # qtable
+    # random
+    # dqn
+    # policy gradient
+    # left to right
+
+
+class QTicTacTable(Agent):
+    def wrapper(self, env):
+        return QtableWrapper(env)
+
+    def __init__(self, alpha=0.1, gamma=0.9):
+        """
+        this will contain the state
+        qdict[state][action] = qvalue
+        where state is a string representing the board and action is an integer
+        """
+        # choice to prune or build
+        # we choose prune here since it's less likely to go wrong
+        super().__init__()
+        self.qdict: Dict[str, Dict[int, float]] = collections.defaultdict(
+            lambda: {i: 0 for i in range(9)}
+        )
+        # update decay - smaller alphas have longer memory, larger alphas shorter
+        self.alpha = alpha
+        # discount factor when updating reward - high gamma is more greedy, low gamma is more patient
+        self.discount = gamma
 
-    def train(self, num_epsiodes, initeps=1, finaleps=0.05):
+    def set_reward(self, state: str, action: int, value: float):
         """
-        Simple linear drop for epsilon.
+        Only allowed to set a state, value pair.
+        Update is as per alpha value
         """
-        epsdecay = (initeps - finaleps) / num_epsiodes
-        epsilon = initeps
+        if self.eval:
+            # no learning in eval mode
+            return
+        self.qdict[state][action] = (
+            self.qdict[state][action] * (1 - self.alpha) + value * self.alpha
+        )
+
+    def prune_qdict(self, state: str, valid_actions: Sequence[int]):
+        """Optional function that allows us to reduce space"""
+        if set(valid_actions) - self.qdict[state].keys():
+            raise ValueError("We have found valid actions not in our qtable for state")
+        elif self.qdict[state].keys() - set(valid_actions):
+            # we can prune down since we don't need the other moves
+            self.qdict[state] = {k: self.qdict[state][k] for k in valid_actions}
+
+    def get_max(self, state: str, valid_actions: Optional[Sequence[int]] = None):
+        """get maximum q score for a state over actions"""
+        if valid_actions:
+            self.prune_qdict(state, valid_actions)
+        return max(self.qdict[state].values())
+
+    def get_arg_max(self, state, valid_actions: Optional[Sequence[int]] = None):
+        """
+        which action gives max q score. If identical max q scrores exist,
+           this'll choose somehing at random
+        """
+        mx = self.get_max(state, valid_actions)
+        # get all moves that return optimal outcome
+        all_arg_maxes = [k for k, v in self.qdict[state].items() if v == mx]
+        return random.choice(all_arg_maxes)
```

Comment:
These two names seem slightly confusing, don't really describe how they relate to one another. Maybe `get_maximum_score` and `get_best_action` or something like that.

Reply:
arg_max and max are very numpy style function names. Max of a list gives you the maximum value while arg_max gives the index where the max value is located.

In this case that's exactly what's happening here. get_max is getting the maximum q value while get_arg_max is getting the action that gives the max value

Reply:
Ok cool, I'm just not numpy-savvy enough to realise this!  But let's run with it, I'll learn something useful.